src/disposable/AsyncDisposable.ts

```diff
@@ -0,0 +1,42 @@
+import { IAsyncDisposable } from './types';
+
+export function createAsyncDisposable(func: () => void | Promise<void>): IAsyncDisposable {
+  return {
+    dispose: func,
+  };
+}
+
+export class AsyncDisposableCollection implements IAsyncDisposable {
+  protected readonly disposables: IAsyncDisposable[] = [];
+
+  get disposed(): boolean {
+    return this.disposables.length === 0;
+  }
+
+  public async dispose(): Promise<void> {
+    if (this.disposed) return;
+
+    while (!this.disposed) {
+      await this.disposables[this.disposables.length - 1].dispose();
+    }
```

Comment:
Ah nooo. I just realized this creates a real bottleneck. @P0lip do we want this to be an `AsyncSeriesDisposable` or and `AsyncConcurrentDisposable`?

Reply:
we might want... both. It depends on the scenario.

Reply:
I went with a "concurrent, but calls them in order, and still handles the case where new disposables are added during the `dispose` phase".

https://github.com/stoplightio/lifecycle/pull/14/files#diff-4457a91706c5e0e61ebd9b792c9f9d06R11-R13

🤔 I should change that to `Promise.allSettled` shouldn't I.

Reply:
> 🤔 I should change that to Promise.allSettled shouldn't I.

Oh god why is that so difficult. OK screw it, it's probably not worth updating TypeScript, adding polyfills, etc etc etc.